pkg/pushsync/pushsync_test.go

```diff
@@ -83,148 +73,191 @@ func TestSendToClosest(t *testing.T) {
 
-	targetCalled := false
-	var mtx sync.Mutex
+	// Create the closest peer
+	psClosestPeer, closestStorerPeerDB := createPushSyncNode(t, closestPeer, nil, mock.WithClosestPeerErr(topology.ErrWantSelf))
+	defer closestStorerPeerDB.Close()
+	defer psClosestPeer.Close()
 
-	// setup the stream recorder to record stream data
-	recorder := streamtest.New(
+	closestRecorder := streamtest.New(
+		streamtest.WithProtocols(psClosestPeer.Protocol()),
 		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
-
-			// this is a custom middleware that is needed because of the design of
-			// the recorder. since we want to test only one unit, but the first message
-			// is supposedly coming from another node, we don't want to execute the handler
-			// when the peer address is the peer of `closestPeer`, since this will create an
-			// unnecessary entry in the recorder
-			return func(ctx context.Context, p p2p.Peer, s p2p.Stream) error {
-				if p.Address.Equal(closestPeer) {
-					mtx.Lock()
-					defer mtx.Unlock()
-					if targetCalled {
-						t.Fatal("target called more than once")
-					}
-					targetCalled = true
-					return nil
-				}
-				return f(ctx, p, s)
-			}
+			return f
 		}),
 	)
 
-	ps := pushsync.New(pushsync.Options{
-		Streamer:      recorder,
-		Logger:        logger,
-		ClosestPeerer: mockTopology,
-		Storer:        storer,
-	})
-	defer ps.Close()
+	// creating the pivot peer
+	psPivot, storerPivotDB := createPushSyncNode(t, pivotPeer, closestRecorder, mock.WithClosestPeer(closestPeer))
+	defer storerPivotDB.Close()
+	defer psPivot.Close()
 
-	recorder.SetProtocols(ps.Protocol())
+	pivotRecorder := streamtest.New(
+		streamtest.WithProtocols(psPivot.Protocol()),
+		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
+			return f
+		}),
+	)
 
-	stream, err := recorder.NewStream(context.Background(), pivotNode, nil, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName)
+	// Creating the trigger peer
+	psTriggerPeer, triggerStorerDB := createPushSyncNode(t, triggerPeer, pivotRecorder, mock.WithClosestPeer(pivotPeer))
+	defer triggerStorerDB.Close()
+	defer psTriggerPeer.Close()
+
+	// upload the chunk to the trigger node DB which triggers the chain reaction of sending this chunk
+	// from trigger peer to pivot peer to closest peer.
+	_, err := triggerStorerDB.Put(context.Background(), storage.ModePutUpload, chunk)
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer stream.Close()
-	w := protobuf.NewWriter(stream)
 
-	// this triggers the handler of the pivot with a delivery stream
-	err = w.WriteMsg(&pb.Delivery{
-		Address: chunkAddress.Bytes(),
-		Data:    chunkData,
-	})
+	// In pivot peer,  intercept the incoming delivery chunk from the trigger peer and check for correctness
+	records := pivotRecorder.WaitRecords(t, pivotPeer, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName, 1, 5)
```

Comment:
the 10 lines of code from `L160` to `L171` repeat themselves constantly in the test, they could be refactored to make the code more readable